In [1]:
from sklearn.feature_selection import chi2
from sklearn.feature_selection import RFE
from sklearn.linear_model import LogisticRegression
import pickle
import matplotlib.pyplot as plt
from sklearn.ensemble import RandomForestClassifier
from sklearn.naive_bayes import GaussianNB
from sklearn.tree import DecisionTreeClassifier
from sklearn.neighbors import KNeighborsClassifier
from sklearn.svm import SVC
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler

In [2]:
def rfefeature(indep_x,dep_y,n):
    rfelist=[]
    log_model=LogisticRegression(solver="lbfgs")
    Rf=RandomForestClassifier(n_estimators=10,criterion="entropy",random_state=0)
    #NB=gaussionNB()
    DT=DecisionTreeClassifier(criterion="gini",max_features="sqrt",splitter="best",random_state=0)
    svc_model=SVC(kernel="linear",random_state=0)
    log_rfe = RFE(estimator=10, n_features_to_select=n)
    #knn=KNeighborsClassification(n_neighbores=5,metric="minkowski",p=2)
    rfemodellist = [log_model, svc_model, Rf, DT]
    for i in rfemodellist:
        print(i)
        log_rfe=RFE(estimator=i, n_features_to_select=n)
        log_fit=log_rfe.fit(indep_x,dep_y)
        log_rfe_feature=log_fit.transform(indep_x)
        rfelist.append(log_rfe_feature)
    return rfelist    

In [3]:
def split_scalar(indep_x,dep_Y):
    # Split the data into training and testing sets
    x_train,x_test,y_train,y_test=train_test_split(indep_x,dep_Y,test_size=0.30,random_state=0)
    sc = StandardScaler() 
    # the training data using the scaler "sc"
    x_train=sc.fit_transform(x_train)
    #Return training and testing sets to independent variables (x_train and x_test) adhoda training and testing sets for dependent variable (y_train and y_test):
    x_test=sc.transform(x_test)
    return x_train,x_test,y_train,y_test

In [4]:
def cm_prediction(classifier,X_test):
        y_pred = classifier.predict(X_test)
        
        # Making the Confusion Matrix
        from sklearn.metrics import confusion_matrix
        cm = confusion_matrix(y_test, y_pred)
        from sklearn.metrics import accuracy_score 
        from sklearn.metrics import classification_report 
            #from sklearn.metrics import confusion_matrix
            #cm = confusion_matrix(y_test, y_pred)
        Accuracy=accuracy_score(y_test, y_pred )
        report=classification_report(y_test, y_pred)
        return  classifier,Accuracy,report,X_test,y_test,cm

In [5]:
def logistic(X_train,y_train,X_test):       
        # Fitting K-NN to the Training set
        from sklearn.linear_model import LogisticRegression
        classifier = LogisticRegression(random_state = 0)
        classifier.fit(X_train, y_train)
        classifier,Accuracy,report,X_test,y_test,cm=cm_prediction(classifier,X_test)
        return  classifier,Accuracy,report,X_test,y_test,cm 

In [6]:
def svm_linear(X_train,y_train,X_test):              
        from sklearn.svm import SVC
        classifier = SVC(kernel = 'linear', random_state = 0)
        classifier.fit(X_train, y_train)
        classifier,Accuracy,report,X_test,y_test,cm=cm_prediction(classifier,X_test)
        return  classifier,Accuracy,report,X_test,y_test,cm
    

In [7]:
def svm_NL(X_train,y_train,X_test):
                
        from sklearn.svm import SVC
        classifier = SVC(kernel = 'rbf', random_state = 0)
        classifier.fit(X_train, y_train)
        classifier,Accuracy,report,X_test,y_test,cm=cm_prediction(classifier,X_test)
        return  classifier,Accuracy,report,X_test,y_test,cm

In [8]:
def Navie(X_train,y_train,X_test):       
        # Fitting K-NN to the Training set
        from sklearn.naive_bayes import GaussianNB
        classifier = GaussianNB()
        classifier.fit(X_train, y_train)
        classifier,Accuracy,report,X_test,y_test,cm=cm_prediction(classifier,X_test)
        return  classifier,Accuracy,report,X_test,y_test,cm

In [9]:
def knn(X_train,y_train,X_test):           
        # Fitting K-NN to the Training set
        from sklearn.neighbors import KNeighborsClassifier
        classifier = KNeighborsClassifier(n_neighbors = 5, metric = 'minkowski', p = 2)
        classifier.fit(X_train, y_train)
        classifier,Accuracy,report,X_test,y_test,cm=cm_prediction(classifier,X_test)
        return  classifier,Accuracy,report,X_test,y_test,cm

In [10]:
def Decision(X_train,y_train,X_test):        
        # Fitting K-NN to the Training set
        from sklearn.tree import DecisionTreeClassifier
        classifier = DecisionTreeClassifier(criterion = 'entropy', random_state = 0)
        classifier.fit(X_train, y_train)
        classifier,Accuracy,report,X_test,y_test,cm=cm_prediction(classifier,X_test)
        return  classifier,Accuracy,report,X_test,y_test,cm 

In [11]:
def random(X_train,y_train,X_test):        
        # Fitting K-NN to the Training set
        from sklearn.ensemble import RandomForestClassifier
        classifier = RandomForestClassifier(n_estimators = 10, criterion = 'entropy', random_state = 0)
        classifier.fit(X_train, y_train)
        classifier,Accuracy,report,X_test,y_test,cm=cm_prediction(classifier,X_test)
        return  classifier,Accuracy,report,X_test,y_test,cm

In [12]:
def rfe_Classification(acclog, accsvml, accsvmnl, accknn, accnav, accdes, accrf): 
    rfedataframe= pd.DataFrame(index=['Logistic',"SVC","Random","DecisionTree"], columns=['Logistic','SVMl','SVMnl','KNN','Navie','Decision','Random'])
    for number, idex in enumerate(rfedataframe.index):      
         rfedataframe['Logistic'][idex] = acclog[number]       
         rfedataframe['SVMl'][idex] = accsvml[number]
         rfedataframe['SVMnl'][idex] = accsvmnl[number]
         rfedataframe['KNN'][idex] = accknn[number]
         rfedataframe['Navie'][idex] = accnav[number]
         rfedataframe['Decision'][idex] = accdes[number]
         rfedataframe['Random'][idex] = accrf[number]
    return rfedataframe

In [13]:
# Load dataset
dataset1=pd.read_csv("Car_Rating data.csv",index_col=None)

In [14]:
df2=dataset1

In [15]:
df2 = pd.get_dummies(df2, drop_first=True)

In [16]:
df2

Index  Overall Rating  Exterior  Interior  Ride Quality  Airbags  \
0        0             4.4       4.4       4.4           4.4       12   
1        1             4.2       4.0       3.8           4.2        8   
2        2             3.5       2.7       3.0           2.3        2   
3        3             4.4       4.4       4.4           4.4        0   
4        4             4.2       4.0       3.8           4.2        4   
..     ...             ...       ...       ...           ...      ...   
167    167             0.0       NaN       NaN           NaN        4   
168    168             0.0       NaN       NaN           NaN        8   
169    169             0.0       NaN       NaN           NaN        4   
170    170             0.0       NaN       NaN           NaN       12   
171    171             0.0       NaN       NaN           NaN        0   

     Price.1  Car_Audi A4  Car_Audi Q3  Car_BMW 2 Series Gran Coupe  ...  \
0      13328        False        False                        False  ...   
1      16621        False        False                        False  ...   
2       8467        False        False                        False  ...   
3       3607        False        False                        False  ...   
4      11726        False        False                        False  ...   
..       ...          ...          ...                          ...  ...   
167    20005        False        False                        False  ...   
168    36110        False        False                        False  ...   
169     7840        False        False                        False  ...   
170      470        False        False                        False  ...   
171     1300        False        False                        False  ...   

     Color_Red  Color_Silver  Color_Sky blue  Color_White  Color_Yellow  \
0        False          True           False        False         False   
1        False         False           False        False         False   
2        False         False           False        False         False   
3        False         False           False         True         False   
4        False          True           False        False         False   
..         ...           ...             ...          ...           ...   
167      False         False           False         True         False   
168      False          True           False        False         False   
169      False         False           False        False         False   
170      False         False           False        False         False   
171      False         False           False        False         False   

     Fuel type_Diesel  Fuel type_Hybrid  Fuel type_LPG  Fuel type_Petrol  \
0               False              True          False             False   
1               False             False          False              True   
2               False             False          False              True   
3               False              True          False             False   
4               False             False          False              True   
..                ...               ...            ...               ...   
167             False             False          False              True   
168             False             False          False              True   
169             False             False          False              True   
170             False             False          False              True   
171             False             False          False              True   

     Fuel type_Plug-in Hybrid  
0                       False  
1                       False  
2                       False  
3                       False  
4                       False  
..                        ...  
167                     False  
168                     False  
169                     False  
170                     False  
171                     False  

[172 rows x 241 

In [17]:
df2["Exterior"].fillna(0, inplace=True)
df2["Interior"].fillna(0, inplace=True)
df2["Ride Quality"].fillna(0, inplace=True)

In [20]:
# Assuming you have already preprocessed the data and have independent and dependent variables
indep_x=df2.drop('Price.1',axis=1)
dep_Y=df2['Price.1']

In [45]:
rfelist=rfefeature(indep_x,dep_Y,1)

LogisticRegression()


C:\Users\ragus\anaconda3\Lib\site-packages\sklearn\linear_model\_logistic.py:458: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(
C:\Users\ragus\anaconda3\Lib\site-packages\sklearn\linear_model\_logistic.py:458: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_

SVC(kernel='linear', random_state=0)
RandomForestClassifier(criterion='entropy', n_estimators=10, random_state=0)
DecisionTreeClassifier(max_features='sqrt', random_state=0)


In [47]:
rfelist=rfefeature(indep_x,dep_Y,1)
acclog=[]
accsvml=[]
accsvmnl=[]
accknn=[]
accnav=[]
accdes=[]
accrf=[]

LogisticRegression()


C:\Users\ragus\anaconda3\Lib\site-packages\sklearn\linear_model\_logistic.py:458: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(
C:\Users\ragus\anaconda3\Lib\site-packages\sklearn\linear_model\_logistic.py:458: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_

SVC(kernel='linear', random_state=0)
RandomForestClassifier(criterion='entropy', n_estimators=10, random_state=0)
DecisionTreeClassifier(max_features='sqrt', random_state=0)


In [48]:
for i in rfelist:
    x_train,x_test,y_train,y_test = split_scalar(i,dep_Y)
    classifier,Accuracy,report,x_test,y_test,cm=logistic(x_train,y_train,x_test)
    acclog.append(Accuracy)
    
    classifier,Accuracy,report,x_test,y_test,cm=svm_linear(x_train,y_train,x_test)  
    accsvml.append(Accuracy)
        
    classifier,Accuracy,report,x_test,y_test,cm=svm_NL(x_train,y_train,x_test)  
    accsvmnl.append(Accuracy)
        
    classifier,Accuracy,report,x_test,y_test,cm=knn(x_train,y_train,x_test)  
    accknn.append(Accuracy)
        
    classifier,Accuracy,report,x_test,y_test,cm=Navie(x_train,y_train,x_test)  
    accnav.append(Accuracy)
        
    classifier,Accuracy,report,x_test,y_test,cm=Decision(x_train,y_train,x_test)  
    accdes.append(Accuracy)
        
    classifier,Accuracy,report,x_test,y_test,cm=random(x_train,y_train,x_test)  
    accrf.append(Accuracy)
        
result=rfe_Classification(acclog,accsvml,accsvmnl,accknn,accnav,accdes,accrf)  

C:\Users\ragus\anaconda3\Lib\site-packages\sklearn\metrics\_classification.py:1344: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
C:\Users\ragus\anaconda3\Lib\site-packages\sklearn\metrics\_classification.py:1344: UndefinedMetricWarning: Recall and F-score are ill-defined and being set to 0.0 in labels with no true samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
C:\Users\ragus\anaconda3\Lib\site-packages\sklearn\metrics\_classification.py:1344: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
C:\Users\ragus\anaconda3\Lib\site-packages\sklearn\metrics\_cla

In [30]:
result
#5

Logistic      SVMl     SVMnl       KNN     Navie  Decision  \
Logistic      0.019231  0.019231  0.019231  0.019231       0.0       0.0   
SVC                0.0       0.0       0.0       0.0       0.0  0.019231   
Random        0.019231  0.019231       0.0  0.038462  0.076923       0.0   
DecisionTree  0.019231  0.019231       0.0  0.038462  0.076923       0.0   

                Random  
Logistic      0.019231  
SVC                0.0  
Random        0.019231  
DecisionTree  0.019231

# Naive Bayes shows the highest performance with a score of 0.076923.
#  Random, Decision Tree, and KNN show moderate performance.
# Logistic and SVC have lower scores.

In [36]:
result
#4

Logistic      SVMl     SVMnl       KNN     Navie Decision  \
Logistic      0.019231  0.019231  0.019231       0.0  0.019231      0.0   
SVC                0.0       0.0       0.0  0.019231       0.0      0.0   
Random        0.019231  0.019231       0.0  0.038462  0.076923      0.0   
DecisionTree  0.019231  0.019231       0.0  0.038462  0.076923      0.0   

                Random  
Logistic      0.019231  
SVC           0.019231  
Random        0.019231  
DecisionTree  0.019231

# Naive Bayes shows the highest performance with a score of 0.076923.
# Random, Decision Tree, and KNN show moderate performance.
# Logistic and SVC have lower scores.

In [40]:
result
#3

Logistic      SVMl     SVMnl       KNN     Navie  Decision  \
Logistic      0.019231  0.019231  0.019231       0.0  0.019231       0.0   
SVC                0.0       0.0       0.0  0.038462       0.0  0.038462   
Random        0.019231  0.019231       0.0  0.038462  0.076923       0.0   
DecisionTree  0.019231       0.0       0.0  0.019231       0.0       0.0   

                Random  
Logistic           0.0  
SVC           0.019231  
Random             0.0  
DecisionTree       0.0

# Naive Bayes shows the highest performance with a score of 0.076923.
# KNN and SVC show moderate performance.
# Logistic, Random, and Decision Tree have lower scores.

In [44]:
result
#2

Logistic SVMl SVMnl       KNN Navie  Decision    Random
Logistic      0.038462  0.0   0.0  0.019231   0.0  0.019231  0.019231
SVC                0.0  0.0   0.0       0.0   0.0       0.0       0.0
Random             0.0  0.0   0.0       0.0   0.0       0.0       0.0
DecisionTree       0.0  0.0   0.0       0.0   0.0       0.0       0.0

# Logistic shows the highest performance with a score of 0.038462.
# KNN, Decision Tree, and Random Forest show lower scores.
# SVC and Random have a score of 0.0.

In [49]:
result
#1

Logistic SVMl SVMnl       KNN     Navie  Decision Random
Logistic          0.0  0.0   0.0  0.019231       0.0  0.019231    0.0
SVC               0.0  0.0   0.0       0.0       0.0       0.0    0.0
Random            0.0  0.0   0.0  0.019231  0.019231       0.0    0.0
DecisionTree      0.0  0.0   0.0  0.019231  0.019231       0.0    0.0